nixos/modules/programs/firefox.nix

```diff
@@ -49,42 +51,106 @@ in {
         ${organisationInfo}
       '';
       type = policyFormat.type;
-      default = {};
+      default = { };
     };
 
     preferences = mkOption {
       description = mdDoc ''
-        Preferences to set from `about://config`.
+        Preferences to set from `about:config`.
 
         Some of these might be able to be configured more ergonomically
         using policies.
 
         ${organisationInfo}
       '';
       type = with types; attrsOf (oneOf [ bool int string ]);
-      default = {};
+      default = { };
+    };
+
+    preferencesStatus = mkOption {
+      description = mdDoc ''
+        The status of `firefox.preferences`.
+
+        `status` can be "default", "locked", "user" or "clear"
+        - `"default"`: Read/Write: Settings appear as default even if factory default differs.
+        - `"locked"`: Read-Only: Settings appear as default even if factory default differs.
+        - `"user"`: Read/Write: Settings appear as changed if it differs from factory default.
+        - `"clear"`: Read/Write: Value has no effect. Resets to factory defaults on each startup.
+      '';
+      type = types.enum [ "default" "locked" "user" "clear" ];
+      default = "locked";
+    };
+
+    autoConfig = mkOption {
+      description = mdDoc ''
+        AutoConfig files can be used to set and lock preferences that are not covered
+        by the policies.json for Mac and Linux. This method can be used to automatically
+        change user preferences or prevent the end user from modifiying specific
+        preferences by locking them.
+      '';
+      type = types.lines;
+      default = "";
+    };
+
+    nativeMessagingHosts = {
+      browserpass = mkEnableOption (lib.mdDoc "Browserpass support");
+      bukubrow = mkEnableOption (lib.mdDoc "Bukubrow support");
+      ff2mpv = mkEnableOption (lib.mdDoc "ff2mpv support");
+      fxCast = mkEnableOption (lib.mdDoc "fx_cast support");
+      gsconnect = mkEnableOption (lib.mdDoc "GSConnect support");
+      jabref = mkEnableOption (lib.mdDoc "JabRef support");
+      passff = mkEnableOption (lib.mdDoc "PassFF support");
+      tridactyl = mkEnableOption (lib.mdDoc "Tridactyl support");
+      ugetIntegrator = mkEnableOption (lib.mdDoc "Uget Integrator support");
     };
   };
 
   config = mkIf cfg.enable {
-    environment.systemPackages = [ cfg.package ];
+    environment.systemPackages = [
+      (cfg.package.override {
+        extraPrefs = cfg.autoConfig;
```

Comment:
Firefox may support read this from etc. https://bugzilla.mozilla.org/show_bug.cgi?id=1170092 We can also apply the patch.

Reply:
According to https://github.com/pyllyukko/user.js#system-wide-installation-all-platforms `/etc/firefox/syspref.js` is supported but I didn't find docs and it seems the api is unstable.

Reply:
It would be good to read this from `/etc` if possible, as that avoids rebuilding the package.

Reply:
AutoConfig in etc requires a patch so we should ask firefox maintains. @lovesegfault @mweinelt

Reply:
I have no interest in this module.

Reply:
Then let's wait upstream. Using the wrapper for now should be good enough.